### PR TITLE
gcs: Overhaul tests and benchmarks.

### PR DESCRIPTION
**Pull request by davecgh**
**This requires PRs #1843 and #1844**.
 
This rewrites the tests to make them more consistent with the rest of the code base and significantly increases their coverage of the code.

It also reworks the benchmarks to actually benchmark what their names claim, renames them for consistency, and make them more stable by ensuring the same prng seed is used each run to eliminate variance introduced by different values.

Finally, it removes an impossible to hit condition from the bit reader and adds a couple of additional checks to harden the filters against potential misuse.

This is part of the ongoing process to cleanup and improve the `gcs` module to the quality level required by consensus code for ultimate inclusion in header commitments.

### CODE REVIEW

**Comment by davecgh** [comment]
```
$ go test -coverprofile=cov.out && go tool cover -func=cov.out
PASS
coverage: 93.8% of statements
ok      github.com/decred/dcrd/gcs/v2   0.422s
github.com/decred/dcrd/gcs/v2/bits.go:18:       writeOne                100.0%
github.com/decred/dcrd/gcs/v2/bits.go:31:       writeZero               100.0%
github.com/decred/dcrd/gcs/v2/bits.go:44:       writeNBits              95.7%
github.com/decred/dcrd/gcs/v2/bits.go:93:       newBitReader            100.0%
github.com/decred/dcrd/gcs/v2/bits.go:102:      readUnary               100.0%
github.com/decred/dcrd/gcs/v2/bits.go:126:      readNBits               97.1%
github.com/decred/dcrd/gcs/v2/gcs.go:43:        Len                     100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:44:        Less                    100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:45:        Swap                    100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:64:        NewFilter               96.9%
github.com/decred/dcrd/gcs/v2/gcs.go:141:       FromBytes               100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:163:       FromNBytes              100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:182:       Bytes                   100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:191:       NBytes                  100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:197:       P                       100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:202:       N                       100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:208:       Match                   100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:250:       MatchAny                100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:314:       readFullUint64          85.7%
github.com/decred/dcrd/gcs/v2/gcs.go:330:       Hash                    100.0%
github.com/decred/dcrd/gcs/v2/gcs.go:346:       MakeHeaderForFilter     0.0%
total:                                          (statements)            93.8%
```

Note that no coverage was added for `MakeHeaderForFilter` because it will removed in upcoming commits.

File: gcs/gcs_test.go [hunk 1]
```diff
@@ -9,218 +7,340 @@ package gcs
...
-	}
-	if f.Match(key, []byte("test")) {
-		t.Fatal("unexpected match of data")
-	}
-	if f.MatchAny(key, nil) {
-		t.Fatal("unexpected match of nil data")
-	}
-	if f.MatchAny(key, [][]byte{[]byte("test")}) {
-		t.Fatal("unexpected match of data")
+// TestFilterCorners ensure a few negative corner cases such as specifying
```
Comment (davecgh) [inline gcs/gcs_test.go]: ensures

File: gcs/gcs_test.go [hunk 2]
```diff
@@ -9,218 +7,340 @@ package gcs
...
 
-	// Ensure empty filters do not have any serialization.
-	gotBytes := f.Bytes()
-	if gotBytes != nil {
-		t.Fatalf("filter bytes not empty -- got %x", gotBytes)
-	}
-	gotBytes = f.NBytes()
-	if gotBytes != nil {
-		t.Fatalf("filter nbytes not empty -- got %x", gotBytes)
+	// Since the filter may have false positives, tries several queries and
```
Comment (davecgh) [inline gcs/gcs_test.go]: try

File: gcs/gcs_test.go [hunk 3]
```diff
@@ -9,218 +7,340 @@ package gcs
...
-	}
-	if f.MatchAny(key, nil) {
-		t.Fatal("unexpected match of nil data")
-	}
-	if f.MatchAny(key, [][]byte{[]byte("test")}) {
-		t.Fatal("unexpected match of data")
+// TestFilterCorners ensure a few negative corner cases such as specifying
+// parameters that are too large behave as expected.
+func TestFilterCorners(t *testing.T) {
+	// Attempt to construct filer with parameters too large.
```
Comment (davecgh) [inline gcs/gcs_test.go]: filter

File: gcs/bench_test.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+var (
+	// globalMatch is used to ensure the benchmarks do not elide code.
+	globalMatch bool
+
+	// Collision probability for the benchmarks (1/2**20).
+	P = uint8(20)
+)
+
+// genFilterElements generates the given number of elements using the provided
+// prng.  This allows prng with a fixed seed to be provided so the same values
```
Comment (davecgh) [inline gcs/bench_test.go]: allows *a*

File: gcs/bench_test.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+// BenchmarkFilterMatchAny benchmarks querying a filter for a list of values.
+func BenchmarkFilterMatchAny(b *testing.B) {
+	// Generate elements for filter.
+	prng1 := rand.New(rand.NewSource(0))
+	contents, err := genFilterElements(20, prng1)
+	if err != nil {
+		b.Fatalf("unable to generate random item: %v", err)
+	}
+
+	// Generate matches uses a separate prng seed so they're very likely all
```
Comment (davecgh) [inline gcs/bench_test.go]: using

File: gcs/gcs_test.go [hunk 6]
```diff
@@ -9,218 +7,340 @@ package gcs
...
-		t.Logf("False positive match, should be 1 in 2**%d!", P)
-	}
-	if filter.Match(key, []byte("Quentins")) {
-		t.Logf("False positive match, should be 1 in 2**%d!", P)
-	}
-	if filter2.Match(key, []byte("Quentins")) {
-		t.Logf("False positive match, should be 1 in 2**%d!", P)
-	}
-}
+		// Ensure the parameters values are returned properly.
```
Comment (davecgh) [inline gcs/gcs_test.go]: parameter
